feat: Create a compatibility score explainer

RahulNavneeth opened the pull request:
### Issue Reference  
- Resolves #6457  

### Feature Implementation  
- Introduces a stateful widget to explain how the compatibility score is computed.  
- Adds sorting functionality, allowing users to sort by score and importance (user preference).  

### Changes Implemented in the Existing Codebase  
1. Added a suffix widget in `SmoothModalSheet`.  
2. Introduced foreground and background colors for `SmoothModalSheetHeaderButton`.  

### Screenshots
<img src="https://github.com/user-attachments/assets/921513cd-0569-411c-91fa-5bed50c23dcb" width="400">  
<img src="https://github.com/user-attachments/assets/c167caf2-2882-4cdb-a3af-6e1f014aa03b" width="400">  

### Video  
#### Before  
[Watch here](https://github.com/user-attachments/assets/6c56482f-c61c-440e-8276-5e024a3c3911)  

#### After  
[Watch here](https://github.com/user-attachments/assets/e1854bde-d894-4edf-b09e-009887b1df42)  

### Potential Improvements
- I couldn't find the exact sort button icon from `SmoothIcon` as specified in the design for #6457 , so I used the `clear` icon as a placeholder. Open to suggestions for a better alternative.

### Additional Notes  
- This PR gave me a deep dive into the codebase, helping me understand how different parts interact.  
- Everything was well-structured, making it easy to jump in and contribute!  
- Super excited to hear your thoughts! 🤗

RahulNavneeth committed:
feat: Create a compatibility score explainer

RahulNavneeth committed:
Merge branch 'develop' into develop

monsieurtanuki commented:
I let @g123k review first the PR, specifically about the UI.

g123k commented:
Hi @RahulNavneeth. Thanks for your PR.

There are some issues with your implementation:
- The banner on top should have a white text
- Sections have a wrong color
- I guess the size of the text in sections is too big
- Text is missing for each attribute
- The bottom buttons are within a container with a slight shadow
- Radius of buttons are incorrect
-…

TLDR: Please implement the design as it's provided in the issue.

RahulNavneeth commented:
Thank you for reviewing @g123k , working on the changes right away 👍

monsieurtanuki commented:
@RahulNavneeth For the record you're supposed to create a branch and to work on it, instead of working on the main `develop` branch.

RahulNavneeth commented:
Understood, @monsieurtanuki 👍. For this particular case, is it okay? I ask because creating a new branch and raising a PR will result in a new PR

monsieurtanuki commented:
Not sure, I hope it's ok.

RahulNavneeth committed:
fix: Compatibility score explainer - Design
Changes made:
1. Updated `smooth_bottom_sheet`'s `foregroundColor` to enforce the parameter-passed value, ensuring the top banner displays white text.
2. Adjusted the section's color and size to match the design.
3. Added titles for all attributes and applied the appropriate compatibility score color.
4. Added a shadow to the footer button container.
5. Modified the footer button's radius.
6. Recreated the SortDown icon as per the design:
   - Combined the sort and down icons.
7. Cleaned up the code.

RahulNavneeth commented:
Hi @g123k , I made the requested design changes. Please review. Thanks!

RahulNavneeth committed:
Merge branch 'develop' into develop

teolemon commented:
/gemini review

g123k commented on packages/smooth_app/lib/generic_lib/bottom_sheets/smooth_bottom_sheet.dart:
```diff
@@ -402,7 +404,7 @@ class SmoothModalSheetHeader extends StatelessWidget implements SizeWidget {
       data: IconThemeData(color: tintColor),
       child: Container(
-        height: suffix is SmoothModalSheetHeaderButton ? double.infinity : null,
+        height: null,
```
Have you checked that you don't break other parts of the app?
This is used for help messages in the Edit mode.

g123k commented on packages/smooth_app/lib/generic_lib/bottom_sheets/smooth_bottom_sheet.dart:
```diff
@@ -527,34 +536,41 @@ class SmoothModalSheetHeaderButton extends StatelessWidget
+            ),
+            child: IconTheme(
+              data: IconThemeData(
+                  color: lightTheme ? Colors.black : Colors.white),
```
Please add a trailing comma

g123k commented on packages/smooth_app/lib/generic_lib/bottom_sheets/smooth_bottom_sheet.dart:
```diff
@@ -527,34 +536,41 @@ class SmoothModalSheetHeaderButton extends StatelessWidget
-                  fontWeight: FontWeight.bold,
-                ),
-                maxLines: 1,
+          child: DefaultTextStyle(
```
Same question: have you checked other modals in the app (settings & edit mode)?

g123k commented on packages/smooth_app/lib/l10n/app_en.arb:
```diff
@@ -4172,6 +4172,16 @@
+  "attribute_score_negative": "Negative points",
+  "attribute_other": "Other",
+  "compatibility_score_modal_title": "Your compatibility score",
+  "sort_by_score": "Sort by : Score",
```
Sort by: Score

g123k commented on packages/smooth_app/lib/l10n/app_en.arb:
```diff
@@ -4172,6 +4172,16 @@
+  "attribute_other": "Other",
+  "compatibility_score_modal_title": "Your compatibility score",
+  "sort_by_score": "Sort by : Score",
+  "sort_by_importance": "Sort by : Importance",
```
Sort by: Importance

g123k commented on packages/smooth_app/lib/l10n/app_en.arb:
```diff
@@ -4172,6 +4172,16 @@
   "@nutrition_facts_extract_failed": {
     "description": "Message to indicate that the extraction of nutrients from a picture failed"
   },
+  "attribute_preference_mandatory": "Mandatory",
+  "attribute_preference_very_important": "Very important",
+  "attribute_preference_important": "Important",
+  "attribute_preference_not_important": "Not important",
```
In the Food preferences screen, `Mandatory, very important…` are values provided by the server.
Please do the same

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+  Attribute.ATTRIBUTE_ALLERGENS_NO_SULPHUR_DIOXIDE_AND_SULPHITES,
+];
+
+class CompatibilityScore extends StatefulWidget {
```
A single Widget for all that stuff is a bad Flutter behavior.
Please split your code into multiple widgets.

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+
+class CompatibilityScore extends StatefulWidget {
+  const CompatibilityScore(
+    this._product, {
```
We don't pass private args this way (cf other parts of the code)

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+    required this.compatibility,
+    this.progress = 0.0,
+    this.maxWidth = 100.0,
+  });
```
Please add some asserts for progress & maxWidth

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+class _CompatibilityScoreState extends State<CompatibilityScore>
+    with UpToDateMixin {
+  bool _sortByImportance = true;
+  final ScrollController _scrollController = ScrollController();
```
Please make the final value, the first one

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+  }
+
+  @override
+  void dispose() {
```
Please move it add the end

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+    );
+  }
+
+  void _showCompatibilityScoreSheet(BuildContext context) {
```
Create a dedicated Widget

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+                      '${widget.compatibility.score ?? AppLocalizations.of(context).not_applicable_short}%',
+                ),
+                body: SizedBox(
+                  height: 500,
```
NO

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+                                final String groupTitle =
+                                    _getGroupTitle(entry.key);
+                                return entry.value.isEmpty
+                                    ? const SizedBox.shrink()
```
EMPTY_WIDGET

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+    }
+  }
+
+  Widget _buildAttributeItem(
```
Create a dedicated Widget

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+    );
+  }
+
+  Widget _getCompatibitlityFooter(StateSetter setModalState) {
```
Create a dedicated Widget + fix your typo

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+      onTap: () => _openFullKnowledgePanel(attribute: attribute),
+      child: Container(
+        width: double.infinity,
+        padding: const EdgeInsets.symmetric(
```
EdgeInsetsDirectional

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+          vertical: SMALL_SPACE,
+          horizontal: MEDIUM_SPACE,
+        ),
+        margin: const EdgeInsets.only(bottom: SMALL_SPACE),
```
EdgeInsetsDirectional

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+        style: OutlinedButton.styleFrom(
+            foregroundColor: foregroundColor, backgroundColor: Colors.white),
+        child: Padding(
+          padding: const EdgeInsets.all(MEDIUM_SPACE),
```
EdgeInsetsDirectional

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+                title: AppLocalizations.of(context)
+                    .compatibility_score_modal_title,
+                prefixIndicator: true,
+                bodyPadding: EdgeInsets.zero,
```
EdgeInsetsDirectional

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+                ),
+                body: SizedBox(
+                  height: 500,
+                  child: SmoothScaffold(
```
Why do you inject a Scaffold for a modal?

g123k commented on packages/smooth_app/lib/pages/product/helpers/compatibility_score.dart:
```diff
@@ -0,0 +1,644 @@
+                  ),
+                ),
+              );
+            }));
```
Add trailing commas

g123k reviewed (changes requested):
Sorry for my late review.
Here are some first comments to fix.

RahulNavneeth commented:
Thanks for reviewing @g123k , Working on the changes right away 👍

teolemon commented:
@RahulNavneeth any update on that ? your PR has small merge conflicts.

g123k commented:
Thanks @RahulNavneeth for your PR.
Now, that we have https://github.com/openfoodfacts/smooth-app/pull/7333, we have a better integration of the Compatibility  score.
